openoa/utils/downloader.py

```diff
@@ -0,0 +1,428 @@
+    result = requests.get(url,stream=True)
+    
+    try:
+        result.raise_for_status()
+        
+        chunk_number = 0
+        
+        try:
+            with outfile.open("wb") as f:
+                
+                for chunk in result.iter_content(chunk_size=1024*1024):
+                    
+                    chunk_number = chunk_number + 1
+                    
+                    print(str(chunk_number) + " MB downloaded", end="\r")
+                    
+                    if chunk: # filter out keep-alive new chunks
+                        f.write(chunk)
+                        
+            logger.info(f"Contents of {url} written to {outfile}")
+            
+        except:
+            logger.error(f"Error writing to {outfile}")
+            
+           
+    except:
+        logger.error(f"Requests.get() returned an error code: {result.status_code}")
+        logger.error(url)
+
+
+def download_zenodo_data(record_id: int, outfile_path: str | Path) -> None:
+    """
+    Download data from zenodo based on the zenodo record_id
+    
+    Args:
+        record_id(:obj:`int`): the Zenodo record id
+        outfile_path(:obj:`str` | :obj:`Path`): path to save files to
+
+    Returns:
+        Files saved to the asset data folder:
+
+          1. record_details.json, which details the zenodo api details
+          2. all files available for the record_id
+    """
+    
+    
+    url_zenodo = r"https://zenodo.org/api/records/"
+    r = requests.get(f"{url_zenodo}{record_id}")
+    
+    r_json = r.json()
+    
+    
+    logger.info("======")
+    logger.info("Title: " + r_json["metadata"]["title"])
+    logger.info("Version: " + r_json["metadata"]["version"])
+    logger.info("URL: " + r_json["links"]["latest_html"])
+    logger.info("Record DOI: " + r_json["doi"])
+    logger.info("License: " + r_json["metadata"]["license"]["id"])
+    logger.info("======\n")
+    
+       
+    # create outfile_path if it does not exist
+    outfile_path = Path(outfile_path).resolve()
+    if not outfile_path.exists():
+        outfile_path.mkdir()
+    
+    
+    # save record details to json file
+    outfile = outfile_path / "record_details.json"
+    
+    with outfile.open("wb") as f:
+        f.write(r.content)
+
+        
+    # download all files
+    files = r_json["files"]
+    for f in files:
+        
+        url_file = f["links"]["self"]
+                
+        outfile = outfile_path / (file_name := f["key"])
+        
+        
+        # check if file exists
+        if outfile.exists():
+            
+            
+            # if it does check the checksum is correct
+            with outfile.open("rb") as f_check:
+                file_hash = hashlib.md5()
+                while chunk := f_check.read(8192):
+                    file_hash.update(chunk)
+        
+            if f["checksum"][4:]==file_hash.hexdigest():
+                logger.info(f"File already exists: {file_name}")
```

Comment:
Is there a way to have this check occur earlier in the process?

Reply:
I don't think so, at least not without slowing things down, as this step compares the file hash of the already downloaded file, with the file hash of the online file. Unfortunately I think the only way to get the hash of the already downloaded file is to read it all. So the process is:

1. Get the json info about the file available online
2. Define the outfile and path to it
3. Check if the outfile already exists, if not jump to step 6
4. If the outfile exists, read the file to get its hash
5. Compare the hash of the file available online (using the json file info) with the hash of the outfile that already exists. If it matches there is no need to download, otherwise move on to step 6
6. Download and unzip the file

Its to make sure if the online file has changed then it is downloaded, even if similar data has already been downloaded.

Reply:
Ok, this makes sense, thanks for adding all this additional info, as it helped a lot to understand the rationale behind the control flow.!